code/weapon/shockwave.cpp

```diff
@@ -297,32 +295,26 @@ void shockwave_move(object *shockwave_objp, float frametime)
 			}
 		}
 
-		// only apply damage to a ship once from a shockwave
-		for ( i = 0; i < sw->num_objs_hit; i++ ) {
-			if ( objp->signature == sw->obj_sig_hitlist[i] ){
-				break;
+		// only apply damage to a on object once from a shockwave if it's on the list
+		for (auto comparison : sw->obj_sig_hitlist) {
+			if ( (objp->signature == comparison.first) && (objp->type == comparison.second) ){
+				continue;
 			}
 		}
 
-		if ( i < sw->num_objs_hit ){
-			continue;
-		}
-
 		if ( weapon_area_calc_damage(objp, &sw->pos, sw->inner_radius, sw->outer_radius, sw->blast, sw->damage, &blast, &damage, sw->radius) == -1 ){
 			continue;
 		}
 
-		// okay, we have damage applied, record the object signature so we don't repeatedly apply damage
-		Assert(sw->num_objs_hit < SW_MAX_OBJS_HIT);
-		if ( sw->num_objs_hit >= SW_MAX_OBJS_HIT) {
-			sw->num_objs_hit--;
-		}
-
 		weapon_info* wip = NULL;
+		
+		// only add non-ships to the list if the Game_settings flag is set
+		if (objp->type == OBJ_SHIP || Shockwaves_consistently_apply_damage){
+		sw->obj_sig_hitlist.push_back({objp->signature, objp->type});
```

Comment:
Use `emplace_back` here and remove the curly braces around the pair.  Also, it looks like the statement needs to be indented.